scripts/ssl/extract_features.py

```diff
@@ -0,0 +1,285 @@
+                " with ``processed_signal`` and ``processed_signal_len`` arguments."
+            )
+        if not has_processed_signal:
+            processed_signal, processed_signal_length = self.preprocessor(
+                input_signal=input_signal,
+                length=input_signal_length,
+            )
+        encoded, encoded_len = self.feature_extractor(audio_signal=processed_signal, length=processed_signal_length)
+        return encoded, encoded_len
+
+
+def maybe_save_features(output_dir, results, max_cache, manifest):
+    """
+    Check if the cache is full and save features to disk
+    """
+    if len(results) == 0 or max_cache < 0 or len(results) < max_cache:
+        return
+    os.makedirs(output_dir, exist_ok=True)
+    logging.info(f"Saving {len(results)} features to {output_dir}")
+
+    for sample_id, audio_file, features_np in tqdm(results, desc="Saving features", total=len(results)):
+        filename = str(audio_file).replace("/", "_").replace(".", "_")
+        if len(filename) > 256:
+            filename = filename[-256:]
+        output_path = os.path.join(output_dir, f"{filename}.npy")
+        np.save(output_path, features_np)
+        manifest[sample_id]["feature_path"] = output_path
+
+    logging.info(f"Saved {len(results)} features to {output_dir}")
+    results.clear()
+
+
+def extract_features(args):
+    """
+    Main function to extract and save features from SSL model
+    """
+
+    logging.info(f"Extracting features using params: {vars(args)}")
+
+    # Load model
+    model = load_model(args.model_path)
+    feature_extractor = FeatureExtractor(model, args.layers)
+    device = torch.device(args.device)
+    feature_extractor.to(device)
+
+    # Load data
+    logging.info(f"Building dataset from input: {args.input}")
+    tmp_manifest = tempfile.NamedTemporaryFile(mode="w", delete=False)
+    manifest = get_input_manifest(args.input)
+    write_manifest(tmp_manifest.name, manifest)
+    total_num_samples = len(manifest)
+
+    # Build dataloader
+    config = {
+        "manifest_filepath": tmp_manifest.name,
+        "sample_rate": feature_extractor.sample_rate,
+        "return_sample_id": True,
+    }
+    dataset = get_char_dataset(config)
+    logging.info(f"Built dataset with {len(dataset)} samples")
+    dataloader = torch.utils.data.DataLoader(
+        dataset=dataset,
+        collate_fn=dataset.collate_fn,
+        batch_size=args.batch_size,
+        shuffle=False,
+        num_workers=args.workers,
+        pin_memory=True,
+        drop_last=False,
+    )
+
+    # Extract features
+    indices = set()
+    results = []
+    amp_dtype = torch.float16 if args.amp_dtype == "float16" else torch.bfloat16
+    logging.info(f"Extracting features using AMP: {args.use_amp}, dtype: {amp_dtype}")
+    with torch.amp.autocast('cuda' if torch.cuda.is_available() else 'cpu', dtype=amp_dtype, enabled=args.use_amp):
```

Comment:
Uh why make it a lightning module if you're manually doing forward pass anyway? Its fine by me, just odd.